File: backend/services/stigmer-server/pkg/domain/workflow/validation/expression_warnings.go
package validation

import (
	"fmt"
	"strings"

	workflowv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/workflow/v1"
	"google.golang.org/protobuf/types/known/structpb"
)

// CheckExpressionWarnings scans all task configs for expressions that reference
// $context.env.* which is a common authoring mistake. Environment variables are
// accessed via $env.*, not $context.env.* ($context holds accumulated exported
// task outputs, not environment variables).
func CheckExpressionWarnings(spec *workflowv1.WorkflowSpec) []string {
	if spec == nil || len(spec.Tasks) == 0 {
		return nil
	}

	var warnings []string

	for _, task := range spec.Tasks {
		config := task.GetTaskConfig()
		if config == nil {
			continue
		}

		hits := findContextEnvRefs(config)
		for _, hit := range hits {
			envKey := hit
			warnings = append(warnings, fmt.Sprintf(
				"task '%s': expression references '$context.env.%s' which resolves to null. "+
					"Environment variables are accessed via '$env.%s', not '$context.env.%s'. "+
					"$context holds accumulated task outputs, not environment variables.",
				task.Name, envKey, envKey, envKey,
			))
		}
	}

	return warnings
}

// findContextEnvRefs scans a Struct value for strings containing "$context.env.".
// Returns the env key names referenced (e.g., "NOTIFICATION_DATE").
func findContextEnvRefs(s *structpb.Struct) []string {
	if s == nil {
		return nil
	}
	var keys []string
	for _, v := range s.Fields {
		keys = append(keys, findContextEnvRefsInValue(v)...)
	}
	return keys
}

func findContextEnvRefsInValue(v *structpb.Value) []string {
	if v == nil {
		return nil
	}

	switch k := v.GetKind().(type) {
	case *structpb.Value_StringValue:
		return extractContextEnvKeys(k.StringValue)
	case *structpb.Value_StructValue:
		return findContextEnvRefs(k.StructValue)
	case *structpb.Value_ListValue:
		var keys []string
		for _, elem := range k.ListValue.GetValues() {
			keys = append(keys, findContextEnvRefsInValue(elem)...)
		}
		return keys
	default:
		return nil
	}
}

// extractContextEnvKeys extracts env key names from expressions like
// "${ $context.env.NOTIFICATION_DATE }".
func extractContextEnvKeys(s string) []string {
	const pattern = "$context.env."
	var keys []string

	for {
		idx := strings.Index(s, pattern)
		if idx < 0 {
			break
		}
		rest := s[idx+len(pattern):]
		keyEnd := 0
		for keyEnd < len(rest) && isEnvKeyChar(rest[keyEnd]) {
			keyEnd++
		}
		if keyEnd > 0 {
			keys = append(keys, rest[:keyEnd])
		}
		s = rest[keyEnd:]
	}

	return keys
}

func isEnvKeyChar(c byte) bool {
	return (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9') || c == '_'
}

File: backend/services/stigmer-server-ts/src/domain/workflow/validation/expression-warnings.ts
/**
 * Expression-mistake warnings — ports
 * pkg/domain/workflow/validation/expression_warnings.go. Scans all task
 * configs for expressions referencing $context.env.* — a common authoring
 * mistake: environment variables are accessed via $env.*, and $context
 * holds accumulated exported task outputs, not environment variables.
 */
import type { JsonObject, JsonValue } from "@bufbuild/protobuf";

import type { WorkflowSpec } from "@stigmer/protos/ai/stigmer/agentic/workflow/v1/spec_pb";

export function checkExpressionWarnings(
  spec: WorkflowSpec | undefined,
): string[] {
  if (spec === undefined || spec.tasks.length === 0) {
    return [];
  }

  const warnings: string[] = [];

  for (const task of spec.tasks) {
    if (task.taskConfig === undefined) {
      continue;
    }

    for (const envKey of findContextEnvRefs(task.taskConfig)) {
      warnings.push(
        `task '${task.name}': expression references '$context.env.${envKey}' which resolves to null. ` +
          `Environment variables are accessed via '$env.${envKey}', not '$context.env.${envKey}'. ` +
          "$context holds accumulated task outputs, not environment variables.",
      );
    }
  }

  return warnings;
}

/**
 * Scans a raw task config for strings containing "$context.env." and
 * returns the env key names referenced (e.g. "NOTIFICATION_DATE"). The
 * stubs surface Struct fields as plain JsonObject, so the walk is over
 * plain JSON — the same tree Go's structpb walk resolves.
 */
function findContextEnvRefs(s: JsonObject): string[] {
  const keys: string[] = [];
  for (const v of Object.values(s)) {
    keys.push(...findContextEnvRefsInValue(v));
  }
  return keys;
}

function findContextEnvRefsInValue(v: JsonValue): string[] {
  if (typeof v === "string") {
    return extractContextEnvKeys(v);
  }
  if (Array.isArray(v)) {
    const keys: string[] = [];
    for (const elem of v) {
      keys.push(...findContextEnvRefsInValue(elem));
    }
    return keys;
  }
  if (typeof v === "object" && v !== null) {
    return findContextEnvRefs(v);
  }
  return [];
}

/**
 * Extracts env key names from expressions like
 * "${ $context.env.NOTIFICATION_DATE }".
 */
function extractContextEnvKeys(s: string): string[] {
  const pattern = "$context.env.";
  const keys: string[] = [];

  for (;;) {
    const idx = s.indexOf(pattern);
    if (idx < 0) {
      break;
    }
    const rest = s.slice(idx + pattern.length);
    let keyEnd = 0;
    while (keyEnd < rest.length && isEnvKeyChar(rest.charCodeAt(keyEnd))) {
      keyEnd++;
    }
    if (keyEnd > 0) {
      keys.push(rest.slice(0, keyEnd));
    }
    s = rest.slice(keyEnd);
  }

  return keys;
}

function isEnvKeyChar(c: number): boolean {
  return (
    (c >= 0x41 && c <= 0x5a) || // A-Z
    (c >= 0x61 && c <= 0x7a) || // a-z
    (c >= 0x30 && c <= 0x39) || // 0-9
    c === 0x5f // _
  );
}
